### PR TITLE
feat: Implement user-defined entity selection strategies in Presidio Structured

### PR DESCRIPTION
**Pull request by miltonsim**
## Change Description

I implemented an additional `selection_strategy` option in the `generate_analysis()` function, allowing users to define their preferred strategy for entity selection. Previously, the function only enabled selection based on the most common entity, which could result in scenarios where a small number of high-confidence entities were overshadowed by a larger number of low-confidence entities. Now, users have the flexibility to choose between three strategies: _most common_, _highest_confidence_ or _mixed_.

I would love to hear to hear feedback and suggestions!

## Issue reference

This PR fixes issue #1316 

## Checklist

- [X] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [X] I have signed the CLA (if required)
- [X] My code includes unit tests
- [X] All unit tests and lint checks pass locally
- [X] My PR contains documentation updates / additions if required

### CODE REVIEW

File: presidio-structured/presidio_structured/analysis_builder.py [hunk 1]
```diff
@@ -159,6 +159,7 @@ def generate_analysis(
         df: DataFrame,
         n: Optional[int] = None,
         language: str = "en",
+        selection_strategy: str = "most_common"
```
Comment (omri374) [inline presidio-structured/presidio_structured/analysis_builder.py:162]: Please add the selection strategies to the docstring (you have that in the inner function, could be copied here as well)

File: presidio-structured/presidio_structured/analysis_builder.py [hunk 2]
```diff
@@ -230,44 +236,133 @@ def _batch_analyze_df(
...
+        :param flat_results: A list of tuples containing index and RecognizerResult
+        objects from the flattened analysis results.
+        :return: A RecognizerResult object selected based on the mixed strategy.
+        """
+        score_aggregator = self._aggregate_scores(flat_results)
+
+        # Check if the highest score is greater than 0.5 and select accordingly
+        highest_score = max(max(scores) for scores in score_aggregator.values()
+                            if scores)
+        if highest_score > 0.5:
```
Comment (omri374) [inline presidio-structured/presidio_structured/analysis_builder.py:336]: Can the 0.5 be a configurable parameter?

**Comment by omri374** [comment]
Thanks @mitonsim! Left a few minor comments

**Comment by omri374** [comment]
Please also add a note to the docs about the strategies: https://github.com/microsoft/presidio/blob/main/docs/structured/index.md

File: presidio-structured/presidio_structured/analysis_builder.py [hunk 3]
```diff
@@ -230,44 +236,133 @@ def _batch_analyze_df(
...
-        flat_results = [
-            (cell_idx, res)
-            for cell_idx, cell_results in enumerate(analyzer_results)
-            for res in cell_results
-        ]
+        flat_results = self._flatten_results(analyzer_results)
 
-        # Count the occurrences of each entity type in different cells
-        type_counter = Counter(res.entity_type for cell_idx, res in flat_results)
+        # Select the entity based on the desired strategy
```
Comment (omri374) [inline presidio-structured/presidio_structured/analysis_builder.py:260]: Please add a validation to the strategy (raise ValueError if the strategy is not supported)

**Comment by miltonsim** [comment]
@omri374, thanks for your feedback! I've incorporated the requested changes.

However, I encountered an issue where my newly added test cases failed in the Azure Pipeline. The assertion assert `key_recognizer_result_map["street"].entity_type == "NON_PII"` is not passing because the 'street' column is incorrectly identified as a location, whereas, in my local environment, it's correctly identified as NON_PII. 

Could you shed some light on this discrepancy and how I can resolve it?

**Comment by omri374** [comment]
@miltonsim, I'm not sure where the discrepancy is coming from. It could be the seed used for sampling, or a different version of the spaCy model. To reduce noise in unit tests, perhaps it would be easier to choose an entity detected using a simpler logic? like a phone number, email, credit card?

**Comment by omri374** [comment]
Hi @miltonsim would you be interested in continuing to work on this? I would suggest to replace the non-deterministic entities with others to make sure tests pass on all environments.

**Comment by miltonsim** [comment]
@omri374 Apologies for the delay. I'll provide an update by 14 March (Tues)

**Comment by omri374** [comment]
Thanks!

**Comment by miltonsim** [comment]
@omri374 Thank you for waiting. I figured that removing the street column might be the best solution. Please help me run the pipeline to check if it passes!

File: presidio-structured/presidio_structured/analysis_builder.py [hunk 4]
```diff
@@ -154,18 +154,26 @@ class TabularAnalysisBuilder(AnalysisBuilder):
 class PandasAnalysisBuilder(TabularAnalysisBuilder):
     """Concrete configuration generator for tabular data."""
 
+    entity_selection_strategies = {"highest_confidence", "mixed", "most_common"}
+
     def generate_analysis(
```
Comment (Jakob-98) [inline presidio-structured/presidio_structured/analysis_builder.py:159]: Overall, PR looks good to me, nice addition!

One thing which might be nice to address (although you didn't introduce this issue) is that we are redefining the function definition of generate_analysis from the AnalysisBuilder(ABC)

Perhaps we can address it in this PR as well, and redefine the abstract method in the base class to: 

`
    @abstractmethod
    def generate_analysis(
        self,
        data: Union[Dict, DataFrame],
        *args,
        **kwargs
    ) -> StructuredAnalysis:
`

(.. and then rename df to data to be consistent)

So we don't override the function definition in the concrete implementations of the class.

**Review by Jakob-98 — APPROVED** [review]

File: presidio-structured/presidio_structured/analysis_builder.py (hunk 4 above)
Comment (omri374) [reply presidio-structured/presidio_structured/analysis_builder.py:159]: I suggest to handle this in a subsequent PR if that's ok with everyone.

**Review by omri374 — APPROVED** [review]

**Comment by elbud** [comment]
Could this update be pushed to PyPi? There is only the original release available at the moment. https://pypi.org/project/presidio-structured/#history

**Comment by omri374** [comment]
Hi @elbud, we plan to have a new release in the next couple of weeks.

**Pull request merged**